coral/breakpoint/breakpoint_graph.py

```diff
@@ -0,0 +1,1003 @@
+        """Add a new node to the breakpoint graph.
+
+        Args:
+                node_: A breakpoint node.
+
+        """
+        if type(node_) != tuple or len(node_) != 3:
+            raise Exception("Breakpoint node must be of form (chr, pos, orientation).")
+        if node_ in self.nodes:
+            pass
+        self.nodes[node_] = [[], [], [], []]
+
+    def add_endnode(self, node_):
+        """Add a new node to the list corresponding to interval ends.
+
+        Args:
+                node_: A breakpoint node.
+
+        """
+        if type(node_) != tuple or len(node_) != 3:
+            raise Exception("Breakpoint node must be of form (chr, pos, orientation).")
+        if node_ not in self.endnodes:
+            self.endnodes[node_] = []
+        else:
+            warnings.warn("Node corresponding to interval end already exists.")
+
+    def del_endnode(self, node_):
+        """Delete a node corresponding to interval ends.
+
+        Args:
+                node_: A breakpoint node.
+
+        """
+        if node_ in self.endnodes:
+            del self.endnodes[node_]
+        else:
+            warnings.warn("Node corresponding to interval end not exists.")
+
+    def del_discordant_endnodes(self):
+        """Delete nodes correspond to interval ends and connect to a discordant edges."""
+        del_list = []
+        for node in self.endnodes:
+            if len(self.endnodes[node]) > 0:
+                del_list.append(node)
+        for node in del_list:
+            del self.endnodes[node]
+
+    def add_sequence_edge(self, chr, l, r, sr_count=-1, sr_flag="d", lr_count=-1, lr_nc=0, cn=0.0):
+        """Add a sequence edge to the graph."""
+        if (chr, l, "-") not in self.nodes or (chr, r, "+") not in self.nodes:
+            raise Exception("Breakpoint node must be added first.")
+        lseq = len(self.sequence_edges)
+        self.nodes[(chr, l, "-")][0].append(lseq)
+        self.nodes[(chr, r, "+")][0].append(lseq)
+        self.sequence_edges.append([chr, l, r, sr_count, sr_flag, lr_count, lr_nc, r - l + 1, cn])
+
+    def add_concordant_edge(
+        self,
+        chr1,
+        pos1,
+        o1,
+        chr2,
+        pos2,
+        o2,
+        sr_count=-1,
+        sr_flag="d",
+        lr_count=-1,
+        reads=set([]),
+        cn=0.0,
+    ):
+        """Add a concordant edge to the graph."""
+        if chr1 != chr2 or pos2 != pos1 + 1 or o1 != "+" or o2 != "-":
+            raise Exception("Invalid concordant edge.")
+        if (chr1, pos1, o1) not in self.nodes or (chr2, pos2, o2) not in self.nodes:
+            raise Exception("Breakpoint node must be added first.")
+        lc = len(self.concordant_edges)
+        self.nodes[(chr1, pos1, o1)][1].append(lc)
+        self.nodes[(chr2, pos2, o2)][1].append(lc)
+        self.concordant_edges.append(
+            [chr1, pos1, o1, chr2, pos2, o2, sr_count, sr_flag, lr_count, reads, cn],
+        )
+
+    def add_discordant_edge(
```

Comment:
I think I would like to clean up these edge definitions further and it would be more efficient to update their docstrings at that time - I've made an issue to track this: https://github.com/AmpliconSuite/CoRAL/issues/22.

Reply:
Sounds great.